Fix crash in theme painting on macOS if GPU is not available

pvollan opened the pull request:
#### 829dab614cc143c1b8c69f6b2535f44f254ad932
<pre>
Fix crash in theme painting on macOS if GPU is not available
<a href="https://bugs.webkit.org/show_bug.cgi?id=247327">https://bugs.webkit.org/show_bug.cgi?id=247327</a>
rdar://100386989

Reviewed by Geoffrey Garen.

This is a fix for a theme painting crash when Metal is unavailable and we&apos;re falling back to OpenGL. The fallback is using CVMS, which is
performing JIT&apos;ing, but only JSC is allowed access to the JIT region in the WebContent process. This change blocks access to CVMS in the
sandbox. I have been able to disable Metal and force software GL in the debugger, and have confirmed that we do not crash with this change.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::gpuMachServices):
* Source/WebKit/WebProcess/com.apple.WebProcess.sb.in:

Canonical link: <a href="https://commits.webkit.org/256539@main">https://commits.webkit.org/256539@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/61d437d73e1e517785ead5bfbe3938c9539ac872

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/96088 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/5337 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/29134 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/105639 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/5456 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/34098 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/88467 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/101457 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/101748 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/4038 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/82694 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/31063 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/85879 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/87788 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/73897 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/39829 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/19316 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/37504 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/20660 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4525 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/42100 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/43265 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/43990 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/39925 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

geoffreygaren commented on Source/WebCore/platform/mac/ThemeMac.mm:
```diff
@@ -949,6 +950,13 @@ static void paintColorWell(ControlStates& controlStates, GraphicsContext& contex
+        RELEASE_LOG_FAULT(Painting, "GPU is not available.");
+        return;
+    }
+#endif
```
Early return seems better than crashing -- but do we have a plan to debug this further? Or is this considered normal for some reason?

pvollan replied:
That is a good point! This is not considered normal, and the current theory is that Metal initialization is failing and we fall back to OpenGL rendering, which is not available. I have reached out to a reproducer of this issue, and requested more diagnostics, in order to better understand the root cause.

Thanks for reviewing!

smfr replied:
Can we call `isGPUAvailable` when IOKit is blocked? It consults Metal devices.

pvollan replied:
> Can we call `isGPUAvailable` when IOKit is blocked? It consults Metal devices.

That is a good point, and will not work if called in the WebContent process.

Thanks for reviewing!

geoffreygaren reviewed (approved):
r=me

smfr reviewed (changes requested):
Unclear how it's OK to just break form control rendering in this context.

geoffreygaren reviewed (approved):
r=me

pvollan committed:
Fix crash in theme painting on macOS if GPU is not available
https://bugs.webkit.org/show_bug.cgi?id=247327
rdar://100386989

Reviewed by Geoffrey Garen.

This is a fix for a theme painting crash when Metal is unavailable and we're falling back to OpenGL. The fallback is using CVMS, which is
performing JIT'ing, but only JSC is allowed access to the JIT region in the WebContent process. This change blocks access to CVMS in the
sandbox. I have been able to disable Metal and force software GL in the debugger, and have confirmed that we do not crash with this change.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::gpuMachServices):
* Source/WebKit/WebProcess/com.apple.WebProcess.sb.in:

Canonical link: https://commits.webkit.org/256539@main
